pkg/cli/cli.go

```diff
@@ -248,20 +335,90 @@ func (c cli) buildRootCmd() *cobra.Command {
 	return rootCmd
 }
 
+// filterPluginsByKeys matches plugins known to a cli to a set of plugin keys,
+// returning unambiguously matched plugins.
+// A match occurs if:
+// - Name and version are the same.
+// - Long or short name is the same, and only one version for that name exists.
+// If long or short name is the same, but multiple versions for that name exist,
+// don't guess which version to use and instead return an error.
+func filterPluginsByKeys(versionedPlugins []plugin.Base, cliPluginKeys map[string]string) ([]plugin.Base, error) {
+	if len(cliPluginKeys) == 0 {
+		return versionedPlugins, nil
+	}
+
+	// Find all potential mateches for this plugin's name.
+	definitelyMatch, maybeMatch := []plugin.Base{}, map[string][]plugin.Base{}
+	for pluginName, pluginVersion := range cliPluginKeys {
+		cliPluginKey := plugin.Key(pluginName, pluginVersion)
+		// Prevents duplicate versions with the same name from being added to
+		// maybeMatch.
+		hasDefinitely := false
```

Comment:
The API plugin, for example, can be used N times. So, will not it throw an error when we try to use the API more than once? Should we not store in the Project file the plugin and flags used for it scaffold files?
 
So, I do not think that we can check for duplicates. What is the scenario that we are trying to avoid with this validation?

Reply:
I think that the check verifies that no plugin is specified twice, for example using `go` plugin twice.

Reply:
@camilamacedo86 the API plugin is a child of `v{1,2}.Plugin{}`, which have the same name but different versions. If they supported an intersecting set of project versions, we would have ambiguous names (`go/v1` and `go/v2` would both be viable for those common project versions).

Reply:
Hi @estroz, 

I understand that:
- A plugin allows us to call/exec the command to scaffold the files
- Many kinds of plugins/command can be called/used more than once 
- If a scaffolded be done by a plugin we will track the plugin/version used for it in the project file. ( It is my understanding from the f2f discussion over it and the main reason is to allow us troubleshooting the scenarios)

Then we will have: 
- A plugin/command X which can be used/called (supportable) if the project version is 1 or 2, for example, but was deprecated and removed for v3.
- So, if the project version is == 1 or 2 this plugin can be used otherwise it will throw the unsupportable error when called.

**Let's think about an example:**

- project init with the go/layout plugin
- and then, with 2 GKV (api) scaffolded
 
Then we will have:
```yaml
domain: testproject.org
layout: go/v2.0.0
multigroup: false
repo: sigs.k8s.io/kubebuilder/testdata/project-v2-multigroup
resources:
- group: crew
  kind: Captain
  version: v1
  plugin: api/v2.0.0
- group: crew
  kind: Frigate
  version: v1beta1
  plugin: api/v2.0.0
version: "2"
```

**NOTE** Just the above example, without the interception scenario, shows not be achievable with the current implementation. 

**Regards the interception** 

Let's migrate the above example project to V3. Then, it means that the ` project.version` will be bumped to 3.
So, after we migrate the project let's scaffold a new API using the plugin. 

Then, the project file will be: 

```yaml
domain: testproject.org
layout: go/v2.0.0 -> Indicates the plugin version/used to init the project
multigroup: false
repo: sigs.k8s.io/kubebuilder/testdata/project-v2-multigroup
resources:
- group: crew
  kind: Captain
  version: v1
  plugin: api/v2.0.0 -> Indicates the plugin version/used to scaffold API
- group: crew
  kind: Frigate
  version: v1beta1
  plugin: api/v2.0.0 -> Indicates the plugin version/used to scaffold API
- group: crew
  kind: NewKind
  version: v2
  plugin: api/v3.0.0 -> Indicates that API was scaffolded with an upper plugin version
version: 3"
```

**Who has the responsibility to know if the command/plugin cannot be executed twice?**

IMO it is part of the command context and not of the plugin, which means that if the init (go/layout)  plugin be called more than once the error will come from the command (It is how it is checked/done already)

**See an example:**

- Project init ( init plugin call for example )
```shell
$ mkdir $GOPATH/src/example
$ cd $GOPATH/src/example
$ kubebuilder init --domain my.domain
Writing scaffold for you to edit...
Get controller runtime:
$ go get sigs.k8s.io/controller-runtime@v0.4.0
Update go.mod:
$ go mod tidy
Running make:
$ make
/Users/camilamacedo/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
go build -o bin/manager main.go
Next: define a resource with:
```

Then, let's try to init twice (call the init plugin for the second time)

```shell
$ kubebuilder create api
$ kubebuilder init --domain my.domain
2020/03/05 09:54:36 failed to initialize project: already initialized
```

**Conclision**

- I do not think that duplicates checks for plugins should be implemented 
- The check to allow or not a command/plugin be executed more than once is part of the command context

Is it make sense? 
If not, could you please help me understand what I misunderstand here.

Reply:
I don't think that resources should track layouts because they are not expected to be mixed. That would be like having some scaffold v1 types and some scaffold v2 in the same project. I don't think that is expected to work.

P.S.: I don't see any relation with `hasDefinetely`.

Reply:
Hi @Adirio, 

I understand that:
- The same project can have scaffolded files made from diff origins. For example, by plugins ( via SDK for example) and indeed via Kubuilder too. 
- If we migrate the project from V2 to V2 and the files do not be re-scaffolded with new plugin versions then it means that the version tracked should be kept for we are able to check it. 
- Then, also as described above, my understanding is that we should track all plugin/versions used to do the scaffolds in the order we are able to troubleshoot the scenarios which is something alleged with kubebuilder admins and discussed in our last f2f. 

PS.: It is not about the `hasDefinetely` as described above it is over the checks and validations made to verify if the plugin is duplicated or not (was used or not before) which IMO should not be done.

Reply:
When I talked about duplications in https://github.com/kubernetes-sigs/kubebuilder/pull/1403#discussion_r387598337 I was refering to plugin duplications, having `go/v2` plugin twice in the list of plugins, not if it was used or not before.

But regarding your points, a plugin is not owned by a plugin. Not even a file is owned by a plugin. For example, if you watch the addon plugin that is implemented right now with the old plugin system, it creates some new files apart from the ones created by kubebuilder when you are scaffolding a resource but it also modifies others. So if `go/v2` created `a.go` but `addon/v1` edited it, who owns the file? Who owns the resource? and if `go/v2` created `api/v1/foo_types.go` and `addon/v1` created `api/v1/foo_additional_types.go`, both belonging to the foo v1 resource, who owns the resource? I think that with this demostration it is clear that we cannot assign ownership of neither a resource nor a file to a single plugin.

So the question that remains is, if we can't track ownership, how do we auto-upgrade from scaffolding versions (for example, scaffolding v2 to v3). The answer is easy. There is no auto-upgrade tool. There has never been one. If you take the v1->v2 example, there is a guide in the book that tells the user how to manually upgrade, but once he has follow the guide, the resulting files are the same as if they were scaffolded with v2. So basically to convert a project from one scaffolding version to another you would need to do it manually.

Reply:
Tks for the explanation @Adirio. I got it now.  The duplication checks are just about to check if we call in the implementation of the command which will use the plugin it duplicated.